### ISSUE TITLE
[FB Messenger] Why is the bot not receiving the "messaging_handovers" event from Messenger API?

### ISSUE THREAD

**Issue by etbuilder** [Microsoft/BotBuilder#3764]
## Bot Info 
* SDK Platform: .NET
* SDK Version: 3.11
* Active Channels: Facebook
* Deployment Environment: ngrok

## Issue Description
I am trying to implement Facebook handover protocol to forward the messages to the page by passing the thread control to the page inbox using app id as in docs (263902037430900), the control is passed successfully, however, the bot do not receive any messaging_handovers event even if the app is already subscribed to it.

As we need to know when the event is received to log it, as well as receive the event when the admin mark the conversation as done and pass the control back to the bot.

## Code Example
Follow Facebook guide to enable forward to agent https://developers.facebook.com/docs/messenger-platform/handover-protocol/pass-thread-control#page_inbox

## Reproduction Steps
1. Send call to Facebook to send control to page
2. Wait for event on MessagesController

## Expected Behavior
Receieve an activity with channel data related to webhook

**Comment by JasonSowers** [comment]
Could you please post your code so we can try to reproduce?

**Comment by etbuilder** [comment]
I was able to talk to Facebook about it, and to solve the issue, they told me to listen for the standby event and ignore the messages.

What is happening is after passing the control to the page inbox, the bots becomes the secondary controller and it will receive messages of type standby, and by having a proxy before the botframework connector we can ignore these.

**Comment by JasonSowers** [comment]
Glad you were able to find a solution, good luck!

**Comment by garypretty** [comment]
@etbuilder any chance you can share some sample code for how you implemented this?  I have exactly the same requirement.

**Comment by etbuilder** [comment]
We enabled standy-by Webhook events on our Facebook app, then in our proxy we just check if the message type is standby we ignore it.

The facebook app webhook is connected to our proxy, in the proxy we check raw messages coming from facebook, and based on type we simply forward them to the botconnecter url .

Code in the proxy
`if (IsStandByEvent(webhookRequestJson)) {
                return req.CreateResponse(HttpStatusCode.OK);
            }`

`private static bool IsStandByEvent(dynamic request) {
            try {
                if (request?.entry?[0]?.standby?[0] != null) {
                    return true;
                }
            } catch {
            }

            return false;
        }`

**Comment by garypretty** [comment]
@etbuilder Thanks for that. Exactly what I needed :)

**Comment by etbuilder** [comment]
You are welcome, you can ping me any time :)

**Comment by garypretty** [comment]
Hey.  Have you got an example of the proxy you built between FB and the Bot Connector? Cannot get mine working.

Thanks

Gary

**Comment by etbuilder** [comment]
Sure thing

ours is Http trigger azure function

```
if (req.GetQueryString("hub.mode") == "subscribe" && req.GetQueryString("hub.verify_token") == token) {
                log.Info($"Recieved Validation Request From Facebook with Challenge {Convert.ToInt32(req.GetQueryString("hub.challenge"))}");
                return req.CreateResponse(HttpStatusCode.OK, Convert.ToInt32(req.GetQueryString("hub.challenge")));
            }

jsonContent = await req.Content.ReadAsStringAsync();
webhookRequestJson = JsonConvert.DeserializeObject(jsonContent);
```

Then we clone the request removing x-liveupgrade and return the response
```
// forward request to appropriate webhook
forwardRequest = req.Clone(targetWebhook, log);
response = await RouteRequest(forwardRequest, log);

(!response.IsSuccessStatusCode) {
                        log.Error("response was not successful" + response.StatusCode);
 }

return response;
```

```
public static HttpRequestMessage Clone(this HttpRequestMessage req, string newUri, TraceWriter log) {
            HttpRequestMessage clone = new HttpRequestMessage(req.Method, newUri);

            if (req.Method != HttpMethod.Get) {
                clone.Content = req.Content;
            }
            clone.Version = req.Version;

            foreach (KeyValuePair<string, object> prop in req.Properties) {
                clone.Properties.Add(prop);
            }

            foreach (KeyValuePair<string, IEnumerable<string>> header in req.Headers) {
                if (header.Key == "X-LiveUpgrade") {
                    continue;
                }

                clone.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            //foreach (KeyValuePair<string, IEnumerable<string>> header in clone.Headers)
            //{
            //    log.Info($"header.Key {header.Key}, header.Value {string.Join(",", header.Value.ToArray())}");
            //}

            clone.Headers.Host = new Uri(newUri).Authority;

            return clone;
        }
```

**Comment by garypretty** [comment]
@etbuilder great, thanks! and it is in here that you are checking if the message is a standby message and ignoring if necessary?

**Comment by garypretty** [comment]
@etbuilder Thanks again for the above.  I am struggling to get it compiling. Looks like a few methods are missing, like RouteRequest and it doesn't seem to like GetQueryString either. Would you be able to share a more complete sample?  Sorry to be a pain, just looking forward to getting this working :)

**Comment by etbuilder** [comment]
```
public static string GetQueryString(this HttpRequestMessage request, string key) {
            var queryStrings = request.GetQueryNameValuePairs();
            if (queryStrings == null)
                return null;

            var match = queryStrings.FirstOrDefault(kv => string.Compare(kv.Key, key, true) == 0);
            if (string.IsNullOrEmpty(match.Value))
                return null;

            return match.Value;
        }
```

```
public static async Task<HttpResponseMessage> RouteRequest(HttpRequestMessage forwardRequest, TraceWriter log) {
            HttpClient client = new HttpClient();
            HttpResponseMessage response = await client.SendAsync(forwardRequest);

            // log any failure messages in a table for checking later on
            //
            //
            if (!response.IsSuccessStatusCode) {
                //await InsertDeadRequest(response.ToString(), "backend webhook error", log);
                log.Info($"Failed {response.ToString()}");
            }

            return response;
        }
```

Answering the previous question, the answer is yes.

**Comment by garypretty** [comment]
@etbuilder Thanks for your help. I just got this working perfectly. Really appreciate the directions on this :) I'll blog about it soon and give you a shout out when I do. You got a twitter handle / blog I can use to reference you?

**Comment by garypretty** [comment]
@etbuilder Hey. I just ran a quick test and it looks like the standby messages are not sent to the bot directly, even when the proxy is not in place, as long as you have not subscribed to that webhook in Facebook.  With this being the case, how come you needed the proxy in the first place?

**Comment by pascaly** [comment]
Hello @etbuilder I'm working to a chatbot that needs the operator handover. I'm using as primary my chatbot app and as secondary the InBox of facebook. Anyway I see many strange behaves... It seems that if I operate from the facebook business manager to change the handover status from ARCHIVE (that as far as I understood means chatbot) and InBox (that is the operator) or viceversa, this status is automatically changed by facebook without any apparent reason and without sending any web_hook message. Moreover, the pass_thread message seems to have always the same data, app_id of the main one, both if you pass from main to secondary that from secondary to main. This seems to be a bug.

Basically, handover seems totally unstable.

Did you experienced the same?

Thanks

**Comment by etbuilder** [comment]
Hey,

Once you pass thread control from main to page, are you using the page inbox app id "263902037430900", if the pass is successful you should receive success (I don't think in this case you will receive a web-hook event).

Once you press "Done" on a message from the inbox, you should receive the event, but, the issue is that bot-connector drops it. If you need this event, you need to intercept all calls from the web-hook between Facebook and bot-connector and check if its handover event, forward it directly to your API endpoint.

Further, if you pass thread control to inbox, you will still receive any postback events from bot-connector like button clicks, to resolve this issue, you have to subscribe to standby events, and when you intercept the webhook messages, you can have your logic in which we simply drop them and not forwarding them to the bot-connector.

**Comment by pascaly** [comment]
Hi  @etbuilder, thanks for your reply. In my case, the issue is on facebook it self: even if I disconnect all the webhook and api management for handover, I see that facebook switches from ARCHIVE to INBOX the conversation, automatically. That seems a facebook bug. My question is if someone experienced this behave.

Thanks

**Comment by etbuilder** [comment]
Have you set your configuration for the page, app as primary and page inbox as secondary?

**Comment by pascaly** [comment]
Sure. I see the events and previously it was working. suddenly, it started to switch back to INBOX even if I selected ARCHIVE. Once I select ARCHIVE it correctly moves back to ARCHIVED folder, but after the first message is sent to the BOT it goes back to INBOX. This happens also without BotBuilder.

**Comment by hoangphhut** [comment]
I found that Facebook Messenger support pass thread control from human manager to bot when the conversation was set 'done' by the human. However I don't know the case if a page has several page managers. My experience is that bot receives call back thread control only when ALL page manager set conversation to 'done'. It's not realistic since several page managers but normally only one is active.

Anyone have solution?

**Comment by bhushanvadgave** [comment]
@etbuilder I am looking for an equivalent code sample for NodeJS/Javascript. Can you help?

**Comment by ABinfinity** [comment]
plz anyone help with this similar issue 
https://github.com/Microsoft/BotBuilder/issues/1465

**Comment by 1110770MiguelSilva** [comment]
@garypretty Did you end up doing a blog about this issue? I'm having the same problem, and was seeing if there was a way to solve this without the use of a proxy. I guess the in the bot builder there is no way to know if a message comes after the thread has been passed to the inbox? Like in the MessagesController for example.

**Comment by amitbend** [comment]
@bhushanvadgave @sharmadine 
I created a gist with an example, I made it quick and dirty for now - but it works
https://gist.github.com/amitbend/9c2109603f8913ab46470e3bea3be7c2

**Comment by 1110770MiguelSilva** [comment]
@etbuilder Hello, could you help me build the proxy like you built it? I am trying to build one in this issue: https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/issues/7
